tests/gnrc_dhcpv6_client/tests-as-root/01-run.py

```diff
@@ -8,24 +8,73 @@
 
 import sys
 from testrunner import run
+from ipaddress import (
+    IPv6Address,
+    IPv6Network,
+)
+
+
+IA_NA_ADDRESS_POOL_PREFIX = "2001:db8:1::"
 
 
 def testfunc(child):
     child.expect(r"Iface\s+\d+")
     child.expect(r"inet6 addr:\sfe80:[0-9a-f:]+\s+scope: link")
     child.expect(r"Iface\s+\d+")
     child.expect(r"inet6 addr:\s+fe80:[0-9a-f:]+\s+scope: link")
-    child.expect(r"inet6 addr:\s+(?P<global_addr>[0-9a-f:]+)\s+scope: global")
-    global_addr = child.match.group("global_addr")
+
+    global_addr_1, global_addr_2 = extract_global_addresses(child)
+
+    global_pfx = extract_global_prefix(child)
+
+    test_global_addrs(global_addr_1, global_addr_2, global_pfx)
+
+
+def extract_global_prefix(child):
     child.expect(r"(?P<global_pfx>[0-9a-f:]+)/64\s+dev #\d\s+"
                  r"expires \d+ sec\s+"
                  r"deprecates \d+ sec")
     global_pfx = child.match.group("global_pfx")
+
     if global_pfx.endswith("::"):
         # remove one trailing : in case there are no 0s between prefix and
         # suffix
         global_pfx = global_pfx[0:-1]
-    assert global_addr.startswith(global_pfx)
+
+    return global_pfx
+
+
+def extract_global_address(child):
+    """Expect and extract a global address from the command line."""
+    child.expect(r"inet6 addr:\s+(?P<global_addr>[0-9a-f:]+)\s+scope: global")
+    return child.match.group("global_addr")
+
+
+def extract_global_addresses(child):
+    """Extract two global addresses and return them as a tuple."""
+    return extract_global_address(child), extract_global_address(child)
+
+
+def check_ia_na_addr(ia_na_addr):
+    """Check if the expected IA_NA address has been assigned"""
+    return IPv6Address(ia_na_addr) in IPv6Network("{}/64".format(IA_NA_ADDRESS_POOL_PREFIX))
+
+
+def check_ia_pd_addr(ia_pd_addr, global_pfx):
+    """Check if the expected IA_PD address has been assigned"""
+    return ia_pd_addr.startswith(global_pfx)
+
+
+def check_global_addrs(ia_na_addr, ia_pd_addr, global_pfx):
+    """Perform IA_NA check for the first and IA_PD for the second address"""
+    return check_ia_na_addr(ia_na_addr) and check_ia_pd_addr(ia_pd_addr, global_pfx)
```

Comment:
Due to lazy evaluation, the second function will be skipped when `check_ia_na_addr(ia_na_addr)` returns `False` (so a positive `ia_pd` is never checked). I think this can be prevented by doing the following

```suggestion
    return [i for i, b enumerate([
        check_ia_na_addr(ia_na_addr), 
        check_ia_pd_addr(ia_pd_addr, global_pfx)
    ]) if b]
```

This way an index list of all true checks is returned, with 0 being `check_ia_na_addr()` and 1 being `check_ia_pd_addr()`.

Reply:
You are right, this is actually a blank spot. While your suggestion certainly works, maybe returning a dictionary instead could be a bit easier to read? I think this should also provide a better error message if the assertion fails.

```suggestion
    return {
        "ia_na_check": check_ia_na_addr(ia_na_addr), 
        "ia_pd_check": check_ia_pd_addr(ia_pd_addr, global_pfx),        
    }
```

Reply:
Then however, you need to check if any of the values is true (for my solution too btw which would have been a check if the list is empty). This could be done e.g. with

```python
any(resultX.values())
```

Reply:
However, I like the idea of a dictionary, since it makes interpreting the results with e.g. `pyterm` much easier on a failed assertion.